internal/engine/upper.go

```diff
@@ -143,61 +140,161 @@ func (u Upper) CreateManifests(ctx context.Context, manifests []model.Manifest,
 			select {
 			case <-ctx.Done():
 				return ctx.Err()
-			case event := <-sw.events:
-				if eventContainsConfigFiles(event) {
-					fmt.Println("Event contains config files")
-					newManifest, err := getNewManifestFromTiltfile(ctx, event.manifest.Name)
-					if err != nil {
-						logger.Get(ctx).Infof("getting new manifest error: %v", err)
-						continue
-					}
-					u.buildStates[event.manifest.Name] = BuildStateClean
-					u.manifestOverlay[event.manifest.Name] = newManifest
-				}
-				oldState := u.buildStates[event.manifest.Name]
-				buildState := oldState.NewStateWithFilesChanged(event.files)
-				u.buildStates[event.manifest.Name] = buildState
-
-				spurious, err := buildState.OnlySpuriousChanges()
-				if err != nil {
-					logger.Get(ctx).Infof("build watch error: %v", err)
-				}
-
-				if spurious {
-					// TODO(nick): I think we probably want to log when this happens?
-					continue
-				}
-
-				manifest := u.getMostRecentManifest(event)
-				err = u.buildManifestFromBuildState(ctx, manifest, buildState)
+			case fsEvent := <-sw.events:
+				u.handleFSEvent(ctx, engineState, fsEvent)
+			case completedBuild := <-engineState.completedBuilds:
+				err := u.handleCompletedBuild(ctx, completedBuild, engineState, s)
 				if err != nil {
 					return err
 				}
-				u.manifestOverlay[manifest.Name] = manifest
-
-				output.Get(ctx).Summary(s.Output(ctx, u.resolveLB))
-				output.Get(ctx).Printf("Awaiting changes…")
 			case err := <-sw.errs:
 				return err
-			// HACK(maia): until upper notifies the HUD of updates, just tell the HUD to update every second
-			case <-time.After(time.Second):
-				u.hud.Update(view.View{})
 			}
+			u.dispatch(ctx, engineState)
 		}
 	}
 	return nil
 }
 
-func (u Upper) getMostRecentManifest(e manifestFilesChangedEvent) model.Manifest {
-	if newManifest, ok := u.manifestOverlay[e.manifest.Name]; ok {
-		return newManifest
+func (u Upper) dispatch(ctx context.Context, state *engineState) {
+	if len(state.manifestsToBuild) == 0 || state.currentlyBuilding != "" {
+		return
 	}
 
-	return e.manifest
+	mn := state.manifestsToBuild[0]
+	state.manifestsToBuild = state.manifestsToBuild[1:]
+	state.currentlyBuilding = mn
+	ms := state.manifestStates[mn]
+
+	var buildState BuildState
+	if ms.configIsDirty {
+		newManifest, err := getNewManifestFromTiltfile(ctx, mn)
+		if err != nil {
+			logger.Get(ctx).Infof("getting new manifest error: %v", err)
+			state.currentlyBuilding = ""
+			return
+		}
+		ms.lastBuild = BuildStateClean
+		ms.manifest = newManifest
+		ms.configIsDirty = false
```

Comment:
need to clear pendingFileChanges in this case too

Reply:
hm, yeah...also I was thinking we didn't want to put them on the BuildState since that might tell the BaD that there were only those changes, but we def. want to display to the user that we picked up their changes, so really all of that stuff should be outside the else.